python/pyspark/sql/types.py

```diff
@@ -276,7 +276,15 @@ def toInternal(self, dt: datetime.datetime) -> int:
     def fromInternal(self, ts: int) -> datetime.datetime:
         if ts is not None:
             # using int to avoid precision loss in float
-            return datetime.datetime.fromtimestamp(ts // 1000000).replace(microsecond=ts % 1000000)
+            return (
+                datetime.datetime
+                # Set the time zone to UTC because the TIMESTAMP type stores timestamps
+                # as the number of microseconds from the epoch of 1970-01-01T00:00:00.000000Z
+                # in the UTC time zone.
+                .fromtimestamp(ts // 1000000, tz=datetime.timezone.utc).replace(
```

Comment:
Hmmm ... I believe this isn't technically correct ... it was a naive datetime that is a local time when it's collected before. Now they are datetimes are in UTC.

Reply:
One workaround would be to use TimestampNTZType instead of doing this I guess.

Reply:
> it was a naive datetime that is a local time when it's collected before

Just coping of an offset from the epoch in UTC and saying that it is a local offset is incorrect, I do believe. So, the current impl is wrong definitely. What I did here and semantically the same what we do while converting TimestampType to external Java type - java.time.Instant.

Reply:
I think there're implementation details that we should fix. My take was (see quotes from https://docs.python.org/3/library/datetime.html#aware-and-naive-objects)

> ... Whether a naive object represents Coordinated Universal Time (UTC), local time, or time in some other timezone is purely up to the program ...

With `TimestampType`, naive datetimes are interpreted as a local time. Meaning that `java.time.LocalDateTime`
With `TimestampNTZType`, naive datetimes are interpreted as a UTC time. Meaning that `java.time.Instant`

But reading that codes, SQL side is opposite. My understanding was that `TimestampType` is a local time per Spark's session timezone, and `TimestampNTZType` is a UTC time.

Reply:
> TimestampNTZType is a UTC time.

This is NO time zone timestamp that means local or logical timestamp. For example, 2022-12-28 20:23:31 could be in UTC, America/Los_Angeles and Europe/Amsterdam. The same local timestamp represents 3 different physical timestamps.  So, this type matches precisely to
```python
datetime.datetime(year, month, day, hour, minute, second, microsecond, tzinfo=None)
```

> My understanding was that TimestampType is a local time per Spark's session timezone ...

Yep, this one must be bound to a time zone:
```python
datetime.datetime(year, month, day, hour, minute, second, microsecond, tzinfo=ZoneInfo('US/Pacific'))
```
or
```python
datetime.datetime(year, month, day, hour, minute, second, microsecond, tzinfo=datetime.timezone.utc)
```

Reply:
@HyukjinKwon As possible workaround, the result could be converted to local time zone, and the attached time zone could be removed. For instance:
```python
datetime.datetime
                # Set the time zone to UTC because the TIMESTAMP type stores timestamps
                # as the number of microseconds from the epoch of 1970-01-01T00:00:00.000000Z
                # in the UTC time zone.
                .fromtimestamp(ts // 1000000, tz=datetime.timezone.utc).replace(
                    microsecond=ts % 1000000
                ).astimezone(None).replace(tzinfo=None)
```

but this `ts` should be interpreted as an offset in UTC time zone from the epoch otherwise it is incorrect.

Reply:
^ SGTM. At least, we should keep returning naive datetimes ....